_posts/2020-06-10-connect-kafka-spring-kafka.md

```diff
@@ -0,0 +1,575 @@
+
+## Kafka Vocabulary
+
+Let's look at the key terminologies of Kafka,
+
+- **Producer**: Producer is a client that sends records to the Kafka server to the specified topic.
+- **Consumer**: Consumers are the recipients who receive records from the Kafka server.
+- **Broker**: Brokers can create a Kafka cluster by sharing information using Zookeeper. The broker receives records from producers and consumers fetch records from the broker by the topic, partition, and offset.
+- **Cluster**: Kafka is a distributed system. A Kafka cluster contains multiple brokers sharing the workload.
+- **Topic**: Topic is a category name to which records are published and consumers read from topics.
+- **Partition**: Records published to a topic are spread across Kafka cluster into several partitions. Each partition can be associated with a broker to allow consumers to read from a topic in parallel.
+- **Offset**: Offset is a pointer to the last record that Kafka has already sent to a consumer.
+
+## Configuring a Kafka Client
+
+We should have a Kafka server running on our machine. If you don't have Kafka setup on your system, take a look at [Kafka quickstart guide](https://kafka.apache.org/quickstart). Once we have Kafka server up and running, Kafka client can be easily configured with Spring configuration in java or even quicker with Spring Boot. Let us start by adding `spring-kafka` dependency to our pom.xml
+
+```
+<dependency>
+  <groupId>org.springframework.kafka</groupId>
+  <artifactId>spring-kafka</artifactId>
+  <version>2.5.2.RELEASE</version>
+</dependency>
+```
+
+### Using Java Configuration
+
+Lets us now see how to configure Kafka client using Java Configuration for Spring. For better understanding, we have separated `KafkaProducerConfig` and `KafkaConsumerConfig`.
+
+Let's have a look at producer config first:
+
+```java
+@Configuration
+class KafkaProducerConfig {
+
+    @Value("${io.reflectoring.kafka.bootstrap-servers}")
+    private String bootstrapServers;
+
+    @Bean
+    public Map<String, Object> producerConfigs() {
+        Map<String, Object> props = new HashMap<>();
+        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        return props;
+    }
+
+    @Bean
+    public ProducerFactory<String, String> producerFactory() {
+        return new DefaultKafkaProducerFactory<>(producerConfigs());
+    }
+
+    @Bean
+    public KafkaTemplate<String, String> kafkaTemplate() {
+        return new KafkaTemplate<>(producerFactory());
+    }
+}
+```
+
+The above example shows how to configure the Kafka producer to send records. `ProducerFactory` is responsible for creating Kafka Producer instances. `KafkaTemplate` helps us to send records to their respective topic and also it gives a template for executing high-level operations. We'll see more about `KafkaTemplate` in [Sending Records](#sending-records) section.
```

Comment:
I'd rather you replace "record" with "message" throughout the text. That seems to be the more common word used in Kafka.

Reply:
What you are saying is right, "message" is the common word used by developers. But in Kafka terminology they call it as "record". That was the reason I kept it like that. 

But for readers as you said "message" makes it more clear.